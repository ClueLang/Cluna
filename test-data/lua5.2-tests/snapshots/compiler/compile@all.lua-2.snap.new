---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
