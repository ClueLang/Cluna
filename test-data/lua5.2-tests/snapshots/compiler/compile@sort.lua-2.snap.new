---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/sort.lua
---
print("testing (parts of) table library");
print("testing unpack");
local unpack = table.unpack
local x, y, z, a, n
a = {}
lim = 2000
for i = 1, lim {
    a[(i)] = i
}
assert(select(lim, unpack(a)) == lim && select('#', unpack(a)) == lim);
x = unpack(a)
assert(x == 1);
x = {
    unpack(a)
}
assert(#x == lim && x[(1)] == 1 && x[(lim)] == lim);
x = {
    unpack(a, lim - 2)
}
assert(#x == 3 && x[(1)] == lim - 2 && x[(3)] == lim);
x = {
    unpack(a, 10, 6)
}
assert(next(x) == nil);
x = {
    unpack(a, 11, 10)
}
assert(next(x) == nil);
x, y = unpack(a, 10, 10)
assert(x == 10 && y == nil);
x, y, z = unpack(a, 10, 11)
assert(x == 10 && y == 11 && z == nil);
a, x = unpack({
    1
})
assert(a == 1 && x == nil);
a, x = unpack({
    1, 
    2
}, 1, 1)
assert(a == 1 && x == nil);
if !_no32 {
    assert(!pcall(unpack, {}, 0, 2 ^ 31 - 1));
    assert(!pcall(unpack, {}, 1, 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31), 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31 - 1), 2 ^ 31 - 1));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 0));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 1));
    pcall(unpack, {}, 1, 2 ^ 31);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 1, 2 ^ 31 - 1)
    assert(a == 20 && b == nil);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 2, 2 ^ 31 - 1)
    assert(a == nil && b == 20);
}
print("testing pack");
a = table.pack()
assert(a[(1)] == nil && a.n == 0);
a = table.pack(table)
assert(a[(1)] == table && a.n == 1);
a = table.pack(nil, nil, nil, nil)
assert(a[(1)] == nil && a.n == 4);
print("testing sort");
local fn check(t) {
    local fn f(a, b) {
        assert(a && b);
        return true
    }
    local s, e = pcall(table.sort, t, f)
    assert(!s && e::find("invalid order function"));
}
check({
    1, 
    2, 
    3, 
    4
});
check({
    1, 
    2, 
    3, 
    4, 
    5
});
check({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
global fn check(a, f) {
    f = f || fn (x, y) {
        return x < y
    }
    for n = #a, 2,  - 1 {
        assert(!f(a[(n)], a[(n - 1)]));
    }
}
a = {
    "Jan", 
    "Feb", 
    "Mar", 
    "Apr", 
    "May", 
    "Jun", 
    "Jul", 
    "Aug", 
    "Sep", 
    "Oct", 
    "Nov", 
    "Dec"
}
table.sort(a);
check(a);
global fn perm(s, n) {
    n = n || #s
    if n == 1 {
        local t = {
            unpack(s)
        }
        table.sort(t);
        check(t);
    } else {
        for i = 1, n {
            s[(i)], s[(n)] = s[(n)], s[(i)]
            perm(s, n - 1);
            s[(i)], s[(n)] = s[(n)], s[(i)]
        }
    }
}
perm({});
perm({
    1
});
perm({
    1, 
    2
});
perm({
    1, 
    2, 
    3
});
perm({
    1, 
    2, 
    3, 
    4
});
perm({
    2, 
    2, 
    3, 
    4
});
perm({
    1, 
    2, 
    3, 
    4, 
    5
});
perm({
    1, 
    2, 
    3, 
    3, 
    5
});
perm({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
perm({
    2, 
    2, 
    3, 
    3, 
    5, 
    6
});
limit = 30000
if _soft {
    limit = 5000
}
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
local x = os.clock()
table.sort(a);
print(string.format("Sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
x = os.clock()
table.sort(a);
print(string.format("Re-sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
x = os.clock()
i = 0
table.sort(a, fn (x, y) {
    i = i + 1
    return y < x
});
print(string.format("Invert-sorting other %d elements in %.2f sec., with %i comparisons", limit, os.clock() - x, i));
check(a, fn (x, y) {
    return y < x
});
table.sort({});
for i = 1, limit {
    a[(i)] = false
}
x = os.clock()
table.sort(a, fn (x, y) {
    return nil
});
print(string.format("Sorting %d equal elements in %.2f sec.", limit, os.clock() - x));
check(a, fn (x, y) {
    return nil
});
for i, v with pairs(a) {
    assert(!v || i == 'n' && v == limit);
}
A = {
    "Ã¡lo", 
    "\0first :-)", 
    "alo", 
    "then this one", 
    "45", 
    "and a new"
}
table.sort(A);
check(A);
table.sort(A, fn (x, y) {
    load(string.format("A[%q] = ''", x))();
    collectgarbage();
    return x < y
});
tt = {
    __lt = fn (a, b) {
        return a.val < b.val
    }
}
a = {}
for i = 1, 10 {
    a[(i)] = {
        val = math.random(100)
    }
    setmetatable(a[(i)], tt);
}
table.sort(a);
check(a, tt.__lt);
check(a);
print("OK");
