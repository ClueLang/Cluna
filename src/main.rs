#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
