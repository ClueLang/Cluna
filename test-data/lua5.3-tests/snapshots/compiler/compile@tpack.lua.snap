---
source: src/main.rs
expression: compiled
input_file: test-data/lua5.3-tests/tpack.lua
---
local pack = string.pack
local packsize = string.packsize
local unpack = string.unpack
print("testing pack/unpack");
local NB = 16
local sizeshort = packsize("h")
local sizeint = packsize("i")
local sizelong = packsize("l")
local sizesize_t = packsize("T")
local sizeLI = packsize("j")
local sizefloat = packsize("f")
local sizedouble = packsize("d")
local sizenumber = packsize("n")
local little = (pack("i2", 1) == "\1\0")
local align = packsize("!xXi16")
assert(1 <= sizeshort && sizeshort <= sizeint && sizeint <= sizelong && sizefloat <= sizedouble);
print("platform:");
print(string.format("\tshort %d, int %d, long %d, size_t %d, float %d, double %d,\n\z
   \tlua Integer %d, lua Number %d", sizeshort, sizeint, sizelong, sizesize_t, sizefloat, sizedouble, sizeLI, sizenumber));
print("\t" .. (little && "little" || "big") .. " endian");
print("\talignment: " .. align);
global fn checkerror(msg, f, ...) {
    local status, err = pcall(f, ...)
    assert(!status && string.find(err, msg));
}
assert(unpack("B", pack("B", 0xff)) == 0xff);
assert(unpack("b", pack("b", 0x7f)) == 0x7f);
assert(unpack("b", pack("b", -0x80)) == -0x80);
assert(unpack("H", pack("H", 0xffff)) == 0xffff);
assert(unpack("h", pack("h", 0x7fff)) == 0x7fff);
assert(unpack("h", pack("h", -0x8000)) == -0x8000);
assert(unpack("L", pack("L", 0xffffffff)) == 0xffffffff);
assert(unpack("l", pack("l", 0x7fffffff)) == 0x7fffffff);
assert(unpack("l", pack("l", -0x80000000)) == -0x80000000);
for i = 1, NB {
    local s = string.rep("\xff", i)
    assert(pack("i" .. i, -1) == s);
    assert(packsize("i" .. i) == #s);
    assert(unpack("i" .. i, s) == -1);
    s = "\xAA" .. string.rep("\0", i - 1)
    assert(pack("<I" .. i, 0xAA) == s);
    assert(unpack("<I" .. i, s) == 0xAA);
    assert(pack(">I" .. i, 0xAA) == s::reverse());
    assert(unpack(">I" .. i, s::reverse()) == 0xAA);
}
{
    local lnum = 0x13121110090807060504030201
    local s = pack("<j", lnum)
    assert(unpack("<j", s) == lnum);
    assert(unpack("<i" .. sizeLI + 1, s .. "\0") == lnum);
    assert(unpack("<i" .. sizeLI + 1, s .. "\0") == lnum);
    for i = sizeLI + 1, NB {
        local s = pack("<j", -lnum)
        assert(unpack("<j", s) == -lnum);
        assert(unpack("<i" .. i, s .. ("\xFF")::rep(i - sizeLI)) == -lnum);
        assert(unpack(">i" .. i, ("\xFF")::rep(i - sizeLI) .. s::reverse()) == -lnum);
        assert(unpack("<I" .. i, s .. ("\0")::rep(i - sizeLI)) == -lnum);
        checkerror("does not fit", unpack, "<I" .. i, ("\x00")::rep(i - 1) .. "\1");
        checkerror("does not fit", unpack, ">i" .. i, "\1" .. ("\x00")::rep(i - 1));
    }
}
for i = 1, sizeLI {
    local lstr = "\1\2\3\4\5\6\7\8\9\10\11\12\13"
    local lnum = 0x13121110090807060504030201
    local n = lnum & (~(-1 << (i * 8)))
    local s = string.sub(lstr, 1, i)
    assert(pack("<i" .. i, n) == s);
    assert(pack(">i" .. i, n) == s::reverse());
    assert(unpack(">i" .. i, s::reverse()) == n);
}
{
    local u = 0xf0
    for i = 1, sizeLI - 1 {
        assert(unpack("<i" .. i, "\xf0" .. ("\xff")::rep(i - 1)) == -16);
        assert(unpack(">I" .. i, "\xf0" .. ("\xff")::rep(i - 1)) == u);
        u = u * 256 + 0xff
    }
}
{
    assert(pack(">i2 <i2", 10, 20) == "\0\10\20\0");
    local a, b = unpack("<i2 >i2", "\10\0\0\20")
    assert(a == 10 && b == 20);
    assert(pack("=i4", 2001) == pack("i4", 2001));
}
print("testing invalid formats");
checkerror("out of limits", pack, "i0", 0);
checkerror("out of limits", pack, "i" .. NB + 1, 0);
checkerror("out of limits", pack, "!" .. NB + 1, 0);
checkerror("%(17%) out of limits %[1,16%]", pack, "Xi" .. NB + 1);
checkerror("invalid format option 'r'", pack, "i3r", 0);
checkerror("16%-byte integer", unpack, "i16", string.rep('\3', 16));
checkerror("not power of 2", pack, "!4i3", 0);
checkerror("missing size", pack, "c", "");
checkerror("variable%-length format", packsize, "s");
checkerror("variable%-length format", packsize, "z");
checkerror("invalid format", packsize, "c1" .. string.rep("0", 40));
if packsize("i") == 4 {
    local s = string.rep("c268435456", 2 ^ 3)
    checkerror("too large", packsize, s);
    s = string.rep("c268435456", 2 ^ 3 - 1) .. "c268435455"
    assert(packsize(s) == 0x7fffffff);
}
for i = 1, sizeLI - 1 {
    local umax = (1 << (i * 8)) - 1
    local max = umax >> 1
    local min = ~max
    checkerror("overflow", pack, "<I" .. i, -1);
    checkerror("overflow", pack, "<I" .. i, min);
    checkerror("overflow", pack, ">I" .. i, umax + 1);
    checkerror("overflow", pack, ">i" .. i, umax);
    checkerror("overflow", pack, ">i" .. i, max + 1);
    checkerror("overflow", pack, "<i" .. i, min - 1);
    assert(unpack(">i" .. i, pack(">i" .. i, max)) == max);
    assert(unpack("<i" .. i, pack("<i" .. i, min)) == min);
    assert(unpack(">I" .. i, pack(">I" .. i, umax)) == umax);
}
assert(unpack(">j", pack(">j", math.maxinteger)) == math.maxinteger);
assert(unpack("<j", pack("<j", math.mininteger)) == math.mininteger);
assert(unpack("<J", pack("<j", -1)) == -1);
if little {
    assert(pack("f", 24) == pack("<f", 24));
} else {
    assert(pack("f", 24) == pack(">f", 24));
}
print("testing pack/unpack of floating-point numbers");
for _, n with ipairs({
    0, 
    -1.1, 
    1.9, 
    1 / 0, 
    -1 / 0, 
    1e20, 
    -1e20, 
    0.1, 
    2000.7
}) {
    assert(unpack("n", pack("n", n)) == n);
    assert(unpack("<n", pack("<n", n)) == n);
    assert(unpack(">n", pack(">n", n)) == n);
    assert(pack("<f", n) == pack(">f", n)::reverse());
    assert(pack(">d", n) == pack("<d", n)::reverse());
}
for _, n with ipairs({
    0, 
    -1.5, 
    1 / 0, 
    -1 / 0, 
    1e10, 
    -1e9, 
    0.5, 
    2000.25
}) {
    assert(unpack("<f", pack("<f", n)) == n);
    assert(unpack(">f", pack(">f", n)) == n);
    assert(unpack("<d", pack("<d", n)) == n);
    assert(unpack(">d", pack(">d", n)) == n);
}
print("testing pack/unpack of strings");
{
    local s = string.rep("abc", 1000)
    assert(pack("zB", s, 247) == s .. "\0\xF7");
    local s1, b = unpack("zB", s .. "\0\xF9")
    assert(b == 249 && s1 == s);
    s1 = pack("s", s)
    assert(unpack("s", s1) == s);
    checkerror("does not fit", pack, "s1", s);
    checkerror("contains zeros", pack, "z", "alo\0");
    for i = 2, NB {
        local s1 = pack("s" .. i, s)
        assert(unpack("s" .. i, s1) == s && #s1 == #s + i);
    }
}
{
    local x = pack("s", "alo")
    checkerror("too short", unpack, "s", x::sub(1, -2));
    checkerror("too short", unpack, "c5", "abcd");
    checkerror("out of limits", pack, "s100", "alo");
}
{
    assert(pack("c0", "") == "");
    assert(packsize("c0") == 0);
    assert(unpack("c0", "") == "");
    assert(pack("<! c3", "abc") == "abc");
    assert(packsize("<! c3") == 3);
    assert(pack(">!4 c6", "abcdef") == "abcdef");
    assert(pack("c3", "123") == "123");
    assert(pack("c0", "") == "");
    assert(pack("c8", "123456") == "123456\0\0");
    assert(pack("c88", "") == string.rep("\0", 88));
    assert(pack("c188", "ab") == "ab" .. string.rep("\0", 188 - 2));
    local a, b, c = unpack("!4 z c3", "abcdefghi\0xyz")
    assert(a == "abcdefghi" && b == "xyz" && c == 14);
    checkerror("longer than", pack, "c3", "1234");
}
{
    local x = pack("<b h b f d f n i", 1, 2, 3, 4, 5, 6, 7, 8)
    assert(#x == packsize("<b h b f d f n i"));
    local a, b, c, d, e, f, g, h = unpack("<b h b f d f n i", x)
    assert(a == 1 && b == 2 && c == 3 && d == 4 && e == 5 && f == 6 && g == 7 && h == 8);
}
print("testing alignment");
{
    assert(pack(" < i1 i2 ", 2, 3) == "\2\3\0");
    local x = pack(">!8 b Xh i4 i8 c1 Xi8", -12, 100, 200, "\xEC")
    assert(#x == packsize(">!8 b Xh i4 i8 c1 Xi8"));
    assert(x == "\xf4" .. "\0\0\0" .. "\0\0\0\100" .. "\0\0\0\0\0\0\0\xC8" .. "\xEC" .. "\0\0\0\0\0\0\0");
    local a, b, c, d, pos = unpack(">!8 c1 Xh i4 i8 b Xi8 XI XH", x)
    assert(a == "\xF4" && b == 100 && c == 200 && d == -20 && (pos - 1) == #x);
    x = pack(">!4 c3 c4 c2 z i4 c5 c2 Xi4", "abc", "abcd", "xz", "hello", 5, "world", "xy")
    assert(x == "abcabcdxzhello\0\0\0\0\0\5worldxy\0");
    local a, b, c, d, e, f, g, pos = unpack(">!4 c3 c4 c2 z i4 c5 c2 Xh Xi4", x)
    assert(a == "abc" && b == "abcd" && c == "xz" && d == "hello" && e == 5 && f == "world" && g == "xy" && (pos - 1) % 4 == 0);
    x = pack(" b b Xd b Xb x", 1, 2, 3)
    assert(packsize(" b b Xd b Xb x") == 4);
    assert(x == "\1\2\3\0");
    a, b, c, pos = unpack("bbXdb", x)
    assert(a == 1 && b == 2 && c == 3 && pos == #x);
    assert(packsize("!8 xXi8") == 8);
    local pos = unpack("!8 xXi8", "0123456701234567")
    assert(pos == 9);
    assert(packsize("!8 xXi2") == 2);
    local pos = unpack("!8 xXi2", "0123456701234567")
    assert(pos == 3);
    assert(packsize("!2 xXi2") == 2);
    local pos = unpack("!2 xXi2", "0123456701234567")
    assert(pos == 3);
    assert(packsize("!2 xXi8") == 2);
    local pos = unpack("!2 xXi8", "0123456701234567")
    assert(pos == 3);
    assert(packsize("!16 xXi16") == 16);
    local pos = unpack("!16 xXi16", "0123456701234567")
    assert(pos == 17);
    checkerror("invalid next option", pack, "X");
    checkerror("invalid next option", unpack, "XXi", "");
    checkerror("invalid next option", unpack, "X i", "");
    checkerror("invalid next option", pack, "Xc1");
}
{
    local x = pack("i4i4i4i4", 1, 2, 3, 4)
    for pos = 1, 16, 4 {
        local i, p = unpack("i4", x, pos)
        assert(i == pos /_ 4 + 1 && p == pos + 4);
    }
    for pos = 0, 12 {
        local i, p = unpack("!4 i4", x, pos + 1)
        assert(i == (pos + 3) /_ 4 + 1 && p == i * 4 + 1);
    }
    local i, p = unpack("!4 i4", x, -4)
    assert(i == 4 && p == 17);
    local i, p = unpack("!4 i4", x, -7)
    assert(i == 4 && p == 17);
    local i, p = unpack("!4 i4", x, -#x)
    assert(i == 1 && p == 5);
    for i = 1, #x + 1 {
        assert(unpack("c0", x, i) == "");
    }
    checkerror("out of string", unpack, "c0", x, 0);
    checkerror("out of string", unpack, "c0", x, #x + 2);
    checkerror("out of string", unpack, "c0", x, -(#x + 1));
}
print("OK");
